### PR TITLE
4652 public priceAggregator medians off-chain

### LINKED ISSUE #4652: Expose key Price Aggregator economic data via subscription
This data can be used in Blockchain Explorers and via sites like The Graph for the community to get key information about the state of the RUN Protocol components.

From [our meeting on Feb 22](https://docs.google.com/document/d/1HkNLG8Kmm9DV6XGxvZEGVWUtHli2OfkR5ph7Q77x3QE/edit#), here is the data we want to expose from the Oracle:

1. For each asset pair
    1. Aggregate price https://github.com/Agoric/agoric-sdk/pull/5850
    2. Source information (per oracle node operator) Deferred to https://github.com/Agoric/agoric-sdk/issues/5854
3. ~Governance parameters~ there are none

Output stream of priceAggregator should be the same as its priceAuthority.

### PR DESCRIPTION
**Pull request by turadg**
closes: #4652

## Description

For each priceAggregator instance, create an off-chain storage stream.

Whenever a quote is produced by its priceAuthority, publish it on that strream.

### Security Considerations

Publishes prices, already public. For marshalling uses readonly.

### Documentation Considerations

zoe README updated.

### Testing Considerations

Updated the primary `priceAggregator` test to verify the publications along with other quotes.

To test off-chain I needed a publication, but it doesn't publish on boot. So to test that I hacked in a `.publish()` and verified the value came out at the desired key. Documented in README. Advice and requests welcomed for any other testing needed.

### CODE REVIEW

File: packages/zoe/test/unitTests/contracts/test-priceAggregator.js [hunk 1]
```diff
@@ -239,35 +256,49 @@ test('median aggregator', async t => {
...
     increment: 17n,
   });
+  await publications.nextMatches({ amountOut: 2374n, timestamp: 4n });
+  await publications.nextMatches({ amountOut: 1050n, timestamp: 4n });
 
   const quote4 = await tickAndQuote();
   t.deepEqual(quote4, { amountOut: 1060n, timestamp: 5n });
+  await publications.nextMatches(quote4);
+  await publications.nextMatches({ amountOut: 1060n, timestamp: 5n });
+  await publications.nextMatches({ amountOut: 1060n, timestamp: 5n });
```
Comment (turadg) [inline packages/zoe/test/unitTests/contracts/test-priceAggregator.js]: should the aggregator skip publishing repeats?

File: packages/zoe/src/contracts/priceAggregator.js [hunk 2]
```diff
@@ -239,7 +254,11 @@ const start = async (
     // Publish a new authenticated quote.
     publishedTimestamp = timestamp;
     lastPrice = median;
-    updater.updateState(authenticatedQuote);
+    console.debug(
+      'DEBUG publishing authenticatedQuote amount',
+      authenticatedQuote.quoteAmount.value,
+    );
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/priceAggregator.js]: to be dropped before merge.

File: packages/zoe/test/unitTests/contracts/test-priceAggregator.js [hunk 3]
```diff
@@ -208,57 +255,118 @@ test('median aggregator', /** @param {ExecutionContext} t */ async t => {
...
   await E(price1300Admin).delete();
+  await publications.nextMatches({
+    amountIn: 2n,
+    amountOut: 1975n,
+    timestamp: 8n,
+  });
 
   const quote8 = await tickAndQuote();
   t.deepEqual(quote8, { amountOut: 1001n, timestamp: 9n });
+  await publications.nextMatches({
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/contracts/test-priceAggregator.js]: 1001n (just above)  seems to be wrong.  It gets that result by combining the 9th tick from `price1000` with the 10th tick from `price800`. I don't yet see why that happens.

The result your iterator gets for tick 9 (1985/2)  is correct from my calculations. The 9th tick for `price1000` is 1100, and the 9th tick for `price800` is 885.

Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/test-priceAggregator.js]: Okay, it's not a bug you created or made worse.  Would you mind adding a comment on line 361 referencing #5851?  And you should expect the publication's result to continue to be ` 1985n / 2n `.

**Review by Chris-Hibbert — APPROVED** [review]

File: packages/zoe/src/contracts/priceAggregator.js [hunk 4]
```diff
@@ -63,9 +74,22 @@ const start = async (
     return harden({ quoteAmount, quotePayment });
   };
 
-  const { notifier, updater } = makeNotifierKit();
+  // To notify the price authority of a new median
+  const {
+    notifier: authenticatedQuoteNotifier,
+    updater: authenticatedQuoteUpdater,
```
Comment (turadg) [inline packages/zoe/src/contracts/priceAggregator.js]: maybe instead,
```suggestion
    notifier: medianNotifier,
    updater: medianUpdater,
```

File: packages/zoe/src/contracts/priceAggregator.js [hunk 5]
```diff
@@ -168,13 +192,27 @@ const start = async (
...
       createQuote: makeCreateQuote(),
-      notifier,
+      notifier: authenticatedQuoteNotifier,
       quoteIssuer: quoteKit.issuer,
       timer,
       actualBrandIn: brandIn,
       actualBrandOut: brandOut,
     });
 
+  // for each new quote from the priceAuthority, publish it to off-chain storage
```
Comment (turadg) [inline packages/zoe/src/contracts/priceAggregator.js:200]: reference tech debt https://github.com/Agoric/agoric-sdk/issues/5853

File: packages/vats/src/lib-chainStorage.js [hunk 6]
```diff
@@ -10,6 +10,19 @@ const { details: X } = assert;
 // ASCII alphanumeric plus underscore.
 const pathSegmentPattern = /^[a-zA-Z0-9_-]{1,100}$/;
 
+/**
+ * @param {string} name
+ */
+export const sanitizePathSegment = name => {
+  const candidate = name.replace(/[- ]/g, '_');
```
Comment (gibson042) [inline packages/vats/src/lib-chainStorage.js:17]: @turadg I missed this at the time, but two questions:
1. What is the motivation for replacing `-` (which is a valid path segment character)?
2. Shouldn't this rather specific function be defined in a higher layer?

Comment (turadg) [reply packages/vats/src/lib-chainStorage.js:17]: Let's discuss in https://github.com/Agoric/agoric-sdk/pull/5926

**Pull request merged**